### PR TITLE
Allow registering remark plugins

### PR DESCRIPTION
**Pull request by stefanprobst**
**Summary**

This PR allows registering remark plugins (and presets) to customize how markdown is serialized/deserialized in the richtext editor.

Two usecases which have come up are (i) avoid saving autolinks for mdx-compatibility, and (ii) enforce formatting options.

fixes:  #5516
related: #4601

did not mark #4601 as fixed, because this only affects how the richtext editor serializes to markdown. to enforce formatting for the markdown editor, the plaintext content would probably have to be run through a remark pipeline [here](https://github.com/netlify/netlify-cms/blob/3bd36776d6c02a0669df9a5b11f105efee2d2ca3/packages/netlify-cms-widget-markdown/src/MarkdownControl/RawEditor.js#L100) but that means text would change while typing, which is odd (i also ran into issues where the textarea lost focus while typing when i tried it).

(sorry for the formatting chages, these were done by `yarn format`)

**Test plan**

Added seriailzer tests.

**Checklist**

Please add a `x` inside each checkbox:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] Code is formatted via running `yarn format`.
- [x] Tests are passing via running `yarn test`.
- [x] The status checks are successful (continuous integration). Those can be seen below.

### CODE REVIEW

**Comment by stefanprobst** [comment]
hmm not sure about the failing e2e test - it seems to work fine running locally
Screenshot_20210721_143937

**Comment by erezrokah** [comment]
> hmm not sure about the failing e2e test - it seems to work fine running locally

I think it's unrelated. Some tests are flaky on CI

File: website/content/docs/beta-features.md [hunk 1]
```diff
@@ -647,3 +650,19 @@ content
...
+You can register plugins to customize [`remark`](https://github.com/remarkjs/remark), the library used by the richtext editor for serializing and deserializing markdown.
+
+```js
+import { registerRemarkPlugin } from 'netlify-cms-core';
+
+// register a plugin
+registerRemarkPlugin(plugin);
+
+// provide global settings to all plugins, e.g. for customizing `remark-stringify`
+registerRemarkPlugin({ settings: { bullet: '-' } });
```
Comment (erezrokah) [inline website/content/docs/beta-features.md:665]:
```suggestion
// register a plugin
CMS.registerRemarkPlugin(plugin);

// provide global settings to all plugins, e.g. for customizing `remark-stringify`
CMS.registerRemarkPlugin({ settings: { bullet: '-' } });
```

To match other parts of the docs:
https://www.netlifycms.org/docs/custom-widgets/#registerwidget
https://www.netlifycms.org/docs/beta-features/#registering-to-cms-events

File: packages/netlify-cms-widget-markdown/src/serializers/index.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ import remarkToRehype from 'remark-rehype';
 import rehypeToHtml from 'rehype-stringify';
 import htmlToRehype from 'rehype-parse';
 import rehypeToRemark from 'rehype-remark';
+import registry from 'netlify-cms-core';
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/serializers/index.js:10]: Individually widgets don't have the core as a dependency.

We would need to pass down the plugins as props from here https://github.com/netlify/netlify-cms/blob/3bd36776d6c02a0669df9a5b11f105efee2d2ca3/packages/netlify-cms-core/src/components/Editor/EditorControlPane/Widget.js#L296

File: packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+import { markdownToRemark, remarkToMarkdown } from '..';
+
+describe('registered remark plugins', () => {
+  afterEach(() => {
+    registry._clearRemarkPlugins();
+  });
+
+  function withNetlifyLinks() {
+    return transformer;
+    function transformer(tree) {
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js:13]:
```suggestion
    return function transformer(tree) {
```

File: packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+              ],
+            },
+            {
+              type: 'text',
+              value: ' text with ',
+            },
+            {
+              type: 'link',
+              title: null,
+              url: 'https://netlify.com',
```
Comment (erezrokah) [inline packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js:48]:
```suggestion
              url: 'https://this-value-should-be-replaced.com',
```

as I believe the test passes without registering the plugin

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @stefanprobst for this great PR and for adding tests and docs too!

I've added some comments, please let me know what you think.

File: packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js (hunk 4 above)
Comment (stefanprobst) [reply packages/netlify-cms-widget-markdown/src/serializers/__tests__/remarkPlugins.spec.js:47]: ugh, thanks for catching this!

File: packages/netlify-cms-widget-markdown/src/serializers/index.js (hunk 2 above)
Comment (stefanprobst) [reply packages/netlify-cms-widget-markdown/src/serializers/index.js]: done.

not totally sure about the [`CopyPasteVisual` plugin](https://github.com/netlify/netlify-cms/blob/adf7ad78a83aba2c258f8c1af2486eb2522e42e0/packages/netlify-cms-widget-markdown/src/MarkdownControl/plugins/CopyPasteVisual.js) and if something needs to be done in [`htmlToSlate`](https://github.com/netlify/netlify-cms/blob/adf7ad78a83aba2c258f8c1af2486eb2522e42e0/packages/netlify-cms-widget-markdown/src/serializers/index.js#L181) as well.

**Review by erezrokah — APPROVED** [review]
This looks 🚀  @stefanprobst. I refactored the code a bit to pass the plugins as an option property instead of an additional argument.

File: packages/netlify-cms-widget-markdown/src/serializers/index.js (hunk 2 above)
Comment (erezrokah) [reply packages/netlify-cms-widget-markdown/src/serializers/index.js]: I think we don't need it for `htmlToSlate` and the changes to `CopyPasteVisual` look good

**Pull request merged**